src/lib/homestore_backend/replication_message.hpp

```diff
@@ -11,43 +11,69 @@ namespace homeobject {
 
 VENUM(ReplicationMessageType, uint16_t, CREATE_PG_MSG = 0, CREATE_SHARD_MSG = 1, SEAL_SHARD_MSG = 2, PUT_BLOB_MSG = 3,
       DEL_BLOB_MSG = 4, UNKNOWN_MSG = 5);
+VENUM(SyncMessageType, uint16_t, PG_META = 0, SHARD_META = 1, SHARD_BATCH = 2,  LAST_MSG = 3);
+VENUM(ResyncBlobState, uint8_t, NORMAL = 0, DELETED = 1, CORRUPTED = 2);
 
 // magic num comes from the first 8 bytes of 'echo homeobject_replication | md5sum'
 static constexpr uint64_t HOMEOBJECT_REPLICATION_MAGIC = 0x11153ca24efc8d34;
+// magic num comes from the first 8 bytes of 'echo homeobject_resync | md5sum'
+static constexpr uint64_t HOMEOBJECT_RESYNC_MAGIC = 0xbb6813cb4a339f30;
 static constexpr uint32_t HOMEOBJECT_REPLICATION_PROTOCOL_VERSION_V1 = 0x01;
+static constexpr uint32_t HOMEOBJECT_RESYNC_PROTOCOL_VERSION_V1 = 0x01;
 static constexpr uint32_t init_crc32 = 0;
+static constexpr uint64_t LAST_OBJ_ID =ULLONG_MAX;
+static constexpr uint64_t DEFAULT_MAX_BATCH_SIZE_MB =128;
 
 #pragma pack(1)
-struct ReplicationMessageHeader {
+template<typename Header>
+struct BaseMessageHeader {
```

Comment:
I was wondering if it might be worth considering an alternative approach here. Perhaps we could simplify things by declaring a base class, `BaseMessageHeader`, with the relevant interfaces as `virtual`. Then, the derived classes could `override` these implementations. I'm curious to hear your thoughts on this idea and whether it might offer any benefits in terms of clarity and maintainability.

Reply:
If there are no scenarios where a BaseMessageHeader object is used directly, you might consider declaring getSize() as a pure virtual function to ensure that subclasses must implement it. Just a personal suggestion, feel free to proceed in the way you prefer.

```
class BaseMessageHeader {
    uint32_t calculate_crc() const {
        return crc32_ieee(init_crc32, reinterpret_cast<const unsigned char*>(this), getSize());
    }
    virtual size_t getSize() const = 0; 
};

class DerivedMessageHeader : public BaseMessageHeader {
size_t getSize() const override {
        return sizeof(DerivedMessageHeader);
    }
```

Reply:
We tried that before, it will cause an error when we perform `reinterpret_cast` the header in the receiver side, because when using `sizeof` and filling the buffer, the virtual function table will taken into it as well.